### DISCUSSION TITLE
Is it possible to write migration in JS code?

### DISCUSSION THREAD

**Discussion by buunguyen** [prisma/prisma#10854] [Q&A]
I need to migrate some data and the operations are too complex to do in SQL. I wonder if it's possible code a migration in JS using Prisma API instead? I have looked at Prisma seed, which can't be written in JS, but it only runs during DB reset stage, thus can't be used for migration purpose. Any idea is appreciated.

**Comment by austincrim** [comment]
Hey @buunguyen 👋 ,

If you are referring to a **data migration**, you could write a regular Node script using the Prisma Client and run that, depending on your use case and complexity.

Could you provide me more details about what you are trying to accomplish?

**Reply by buunguyen** [reply]
Hi, thanks for your answer. The use case is I need to introduce a new column in a table and already write SQL migration for it, this is not an issue. However, that column needs to be populated with initial data computed from existing columns and this is operation is too complex to do in SQL. I want to create a migration in JS that is executed by Prisma after the above SQL migration runs. That way, I can easily apply the changes to all environments in a predictable manner.

Do you think this is a valid use case? Or this kind of tasks should be done by a one-off Node script out of the migration workflow?

**Comment by jpaas** [comment]
One possible example is adding a non-null column that is populated based on some other data. This is a 3 step process...
1. Add the nullable column
2. Run the script to populate the column
3. Make the column non-nullable.

Drawing a parallel to Rails, Rails migrations are written in ruby. This allows you write any arbitrary code into the migration.

Another neat feature Rails has is that migrations are reversible. You can write migrations that are bidirectional, symmetrically or asymmetrically, again your choice because its just ruby code.

**Reply by vnenkpet** [reply]
But the problem with this is that you might run this on one environment, add another migrations after that, and then deploy this finalized code with all the other migrations that might affect this to the production environment. It just adds a lot of unnecessary manual overhead for no apparent reason. I wish Prisma just worked like any other ORM in this regard.

**Comment by buunguyen** [comment]
> Drawing a parallel to Rails, Rails migrations are written in ruby. This allows you write any arbitrary code into the migration.
> Another neat feature Rails has is that migrations are reversible.

These sound similar to pretty much every ORM I worked with, including Node options like Sequelize. 

I think it would be great to have these capabilities in Prisma, unless there's already a mechanism in Prisma for creating complex migration code.

**Reply by JakeTheCorn** [reply]
Agreed. 

Migrations are an important way to control the changes to db system. 

Often it is even valuable to perform automated testing on a migration. 

It'd be great if prisma could have feature parity with other tools in this respect

**Comment by chientrm** [comment]
1. Add custom `Migration` tracking table:

```prisma
model Migration {
  id        String   @id
  createdAt DateTime @default(now())
}
```

2. Run migrate

```sh
npx prisma migrate dev
```

3. `migration.ts`

```ts
import { Prisma, PrismaClient } from "@prisma/client";

interface Migration {
  id: string;
  migrate: (
    tx: Omit<
      PrismaClient<
        Prisma.PrismaClientOptions,
        never,
        Prisma.RejectOnNotFound | Prisma.RejectPerOperation | undefined
      >,
      "$connect" | "$disconnect" | "$on" | "$transaction" | "$use"
    >
  ) => Promise<void>;
}

const migrations: Migration[] = [
    {
      id: "<migration-id>",
      migrate: async (tx) => {
        // Migration code
      },
    },
  ],
  migrate = async (prisma: PrismaClient) => {
    for (const { id, migrate } of migrations) {
      const migration = await prisma.migration.findFirst({ where: { id } });
      if (migration === null) {
        console.log(`Migrating ${id}`);
        await prisma.$transaction(async (tx) => {
          await migrate(tx);
          await tx.migration.create({ data: { id } });
        });
        console.log(`Migrated ${id}`);
      }
    }
    return prisma;
  };

export { migrate };
```

4. In your main source:

```ts
import { PrismaClient } from "@prisma/client";
import { migrate } from "./migrations";
const forPrisma = migrate(new PrismaClient());

const foo = async () => {
  const prisma = await forPrisma;
}
```

**Comment by isc30** [comment]
Any update on this? The proposed Solutions arent really a valid option. Ideally prisma generated migration would be a js file that does prismaClient.runRawSql(...) but allows for manual modifications to do more complex operations there
